-Wformat clang/gcc warnings

New compilation warnings on CRAN
https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-fedora-clang/data.table-00check.html
Same warnings, but less, on GCC
```
checking whether package ‘data.table’ can be installed ... [57s/138s] WARNING
Found the following significant warnings:
  assign.c:470:83: warning: format specifies type 'int' but the argument has type 'R_xlen_t' (aka 'long') [-Wformat]
  assign.c:518:22: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  assign.c:794:96: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  forder.c:107:30: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:121:19: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:255:27: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:270:23: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:272:21: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:283:18: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:284:23: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:304:25: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:322:17: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:340:14: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:404:3: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:426:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:428:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:430:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:443:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:445:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:447:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:453:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:455:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:459:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:462:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:465:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:467:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:495:5: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:511:9: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:553:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:599:11: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:706:8: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:720:47: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:725:57: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1045:36: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1154:15: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1261:38: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1265:57: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1270:29: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1302:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1315:29: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1332:7: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  forder.c:1366:9: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  fread.c:1888:54: warning: format specifies type 'int' but the argument has type 'int64_t' (aka 'long') [-Wformat]
  fsort.c:168:102: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fsort.c:245:49: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fsort.c:249:54: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fwrite.c:626:28: warning: format specifies type 'int' but the argument has type 'const void *' [-Wformat]
  fwrite.c:711:68: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fwrite.c:731:15: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  fwrite.c:746:64: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fwrite.c:814:10: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fwrite.c:825:10: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  fwrite.c:915:21: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  gsumm.c:90:38: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  gsumm.c:90:48: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  gsumm.c:90:56: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  gsumm.c:90:67: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  gsumm.c:609:90: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  gsumm.c:612:93: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  gsumm.c:638:92: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  gsumm.c:1411:72: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  ijoin.c:323:89: warning: data argument not used by format string [-Wformat-extra-args]
  init.c:153:139: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:168:71: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:169:77: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:171:83: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:172:106: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:173:105: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:173:119: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:174:81: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:175:79: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:176:87: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:177:77: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:178:79: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:179:79: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:180:81: warning: format specifies type 'int' but the argument has type 'unsigned long' [-Wformat]
  init.c:184:95: warning: format specifies type 'int' but the argument has type 'R_xlen_t' (aka 'long') [-Wformat]
  rbindlist.c:212:37: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  rbindlist.c:233:50: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  rbindlist.c:234:48: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  rbindlist.c:494:31: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  reorder.c:17:115: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  reorder.c:27:136: warning: format specifies type 'int' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
  snprintf.c:217:136: warning: format specifies type 'char *' but the argument has type 'int' [-Wformat]
  subset.c:270:26: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
See the [install log](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-fedora-clang/data.table-00install.html) for details.
used C compiler: ‘clang version 17.0.5’
```

--- a/src/assign.c
+++ b/src/assign.c
@@ -527,7 +527,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
       targetcol = VECTOR_ELT(dt,coln);
     }
     const char *ret = memrecycle(targetcol, rows, 0, targetlen, thisvalue, 0, -1, coln+1, CHAR(STRING_ELT(names, coln)));
-    if (ret) warning(ret);
+    if (ret) warning("%s", ret);
   }
 
   *_Last_updated = numToDo;  // the updates have taken place with no error, so update .Last.updated now

--- a/src/forder.c
+++ b/src/forder.c
@@ -56,7 +56,7 @@ static int *anso = NULL;
 static bool notFirst=false;
 
 static char msg[1001];
-#define STOP(...) do {snprintf(msg, 1000, __VA_ARGS__); cleanup(); error(msg);} while(0)      // http://gcc.gnu.org/onlinedocs/cpp/Swallowing-the-Semicolon.html#Swallowing-the-Semicolon
+#define STOP(...) do {snprintf(msg, 1000, __VA_ARGS__); cleanup(); error("%s", msg);} while(0)      // http://gcc.gnu.org/onlinedocs/cpp/Swallowing-the-Semicolon.html#Swallowing-the-Semicolon
 // use STOP in this file (not error()) to ensure cleanup() is called first
 // snprintf to msg first in case nrow (just as an example) is provided in the message because cleanup() sets nrow to 0
 #undef warning

--- a/src/fwrite.c
+++ b/src/fwrite.c
@@ -736,7 +736,7 @@ void fwriteMain(fwriteMainArgs args)
     }
     if (f==-1) {
       *ch = '\0';
-      DTPRINT(buff);
+      DTPRINT("%s", buff);
       free(buff);
     } else {
       int ret1=0, ret2=0;
@@ -926,7 +926,7 @@ void fwriteMain(fwriteMainArgs args)
           errno=0;
           if (f==-1) {
             *ch='\0';  // standard C string end marker so DTPRINT knows where to stop
-            DTPRINT(myBuff);
+            DTPRINT("%s", myBuff);
           } else if ((args.is_gzip ? WRITE(f, myzBuff, (int)myzbuffUsed)
                                    : WRITE(f, myBuff,  (int)(ch-myBuff))) == -1) {
             failed=true;         // # nocov

--- a/src/rbindlist.c
+++ b/src/rbindlist.c
@@ -208,7 +208,7 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
           const char *str = isString(s) ? CHAR(STRING_ELT(s,w2)) : "";
           snprintf(buff, 1000, _("Column %d ['%s'] of item %d is missing in item %d. Use fill=TRUE to fill with NA (NULL for list columns), or use.names=FALSE to ignore column names.%s"),
                         w2+1, str, i+1, missi+1, extra );
-          if (usenames==TRUE) error(buff);
+          if (usenames==TRUE) error("%s", buff);
           i = LENGTH(l); // break from outer i loop
           break;         // break from inner j loop
         }
@@ -229,8 +229,8 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         }
         const char *o = isNull(opt) ? "message" : CHAR(STRING_ELT(opt,0));
         if      (strcmp(o,"message")==0) { eval(PROTECT(lang2(install("message"),PROTECT(ScalarString(mkChar(buff))))), R_GlobalEnv); UNPROTECT(2); }
-        else if (strcmp(o,"warning")==0) warning(buff);
-        else if (strcmp(o,"error")==0)   error(buff);
+        else if (strcmp(o,"warning")==0) warning("%s", buff);
+        else if (strcmp(o,"error")==0)   error("%s", buff);
         else if (strcmp(o,"none")!=0)    warning(_("options()$datatable.rbindlist.check=='%s' which is not 'message'|'warning'|'error'|'none'. See news item 5 in v1.12.2."), o);
       }
     }
@@ -490,7 +490,7 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
       }
       for (int k=0; k<nLevel; ++k) SET_TRUELENGTH(levelsRaw[k], 0);
       savetl_end();
-      if (warnStr[0]) warning(warnStr);  // now savetl_end() has happened it's safe to call warning (could error if options(warn=2))
+      if (warnStr[0]) warning("%s", warnStr);  // now savetl_end() has happened it's safe to call warning (could error if options(warn=2))
       SEXP levelsSxp;
       setAttrib(target, R_LevelsSymbol, levelsSxp=allocVector(STRSXP, nLevel));
       for (int k=0; k<nLevel; ++k) SET_STRING_ELT(levelsSxp, k, levelsRaw[k]);

--- a/src/subset.c
+++ b/src/subset.c
@@ -281,7 +281,7 @@ SEXP subsetDT(SEXP x, SEXP rows, SEXP cols) { // API change needs update NEWS.md
     SEXP max = PROTECT(ScalarInteger(nrow)); nprotect++;
     rows = PROTECT(convertNegAndZeroIdx(rows, max, ScalarLogical(TRUE), ScalarLogical(TRUE))); nprotect++;
     const char *err = check_idx(rows, nrow, &anyNA, &orderedSubset);
-    if (err!=NULL) error(err);
+    if (err!=NULL) error("%s", err);
   }
 
   if (!isInteger(cols)) error(_("Internal error. Argument '%s' to %s is type '%s' not '%s'"), "cols", "Csubset", type2char(TYPEOF(cols)), "integer"); // # nocov

